src/instruction.rs

```diff
@@ -213,143 +213,239 @@ impl fmt::Display for MemoryReference {
+    pub parameters: Vec<String>,
+    pub matrix: Vec<Vec<Expression>>,
+    pub r#type: GateType,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Declaration {
+    pub name: String,
+    pub size: Vector,
+    pub sharing: Option<String>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Measurement {
+    pub qubit: Qubit,
+    pub target: Option<MemoryReference>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Reset {
+    pub qubit: Option<Qubit>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct CalibrationDefinition(pub Calibration);
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Capture {
+    pub blocking: bool,
+    pub frame: FrameIdentifier,
+    pub memory_reference: MemoryReference,
+    pub waveform: WaveformInvocation,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Delay {
+    pub duration: Expression,
+    pub frame_names: Vec<String>,
+    pub qubits: Vec<Qubit>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Fence {
+    pub qubits: Vec<Qubit>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct FrameDefinition {
+    pub identifier: FrameIdentifier,
+    pub attributes: HashMap<String, AttributeValue>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MeasureCalibrationDefinition {
+    pub qubit: Option<Qubit>,
+    pub parameter: String,
+    pub instructions: Vec<Instruction>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Pragma {
+    pub name: String,
+    pub arguments: Vec<String>,
+    pub data: Option<String>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Pulse {
+    pub blocking: bool,
+    pub frame: FrameIdentifier,
+    pub waveform: WaveformInvocation,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct RawCapture {
+    pub blocking: bool,
+    pub frame: FrameIdentifier,
+    pub duration: Expression,
+    pub memory_reference: MemoryReference,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SetFrequency {
+    pub frame: FrameIdentifier,
+    pub frequency: f64,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SetPhase {
+    pub frame: FrameIdentifier,
+    pub phase: Expression,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SetScale {
+    pub frame: FrameIdentifier,
+    pub scale: Expression,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ShiftFrequency {
+    pub frame: FrameIdentifier,
+    pub frequency: f64,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ShiftPhase {
+    pub frame: FrameIdentifier,
+    pub phase: Expression,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SwapPhases {
+    pub frame_1: FrameIdentifier,
+    pub frame_2: FrameIdentifier,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct WaveformDefinition {
+    pub name: String,
+    pub definition: Waveform,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Arithmetic {
+    pub operator: ArithmeticOperator,
+    pub destination: ArithmeticOperand,
+    pub source: ArithmeticOperand,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Halt;
```

Comment:
An empty type for an empty enum is definitely excessive. If the worry is about future compatibility, changing any one of these structs is a breaking change anyway because everything is pub.